crates/bindings/src/table.rs

```diff
@@ -0,0 +1,414 @@
+/// and the module macroexpansion.
+pub trait Table: TableInternal {
+    /// The type of rows stored in this table.
+    type Row: SpacetimeType + Serialize + DeserializeOwned + 'static;
+
+    /// Returns the number of rows of this table in the TX state,
+    /// i.e. num(committed_state) + num(insert_table) - num(delete_table).
+    ///
+    /// This API is new to modules (though it previously existed in the Rust SDK)
+    /// and will require a new host function in the ABI.
+    fn count(&self) -> u64 {
+        sys::datastore_table_row_count(Self::table_id()).expect("datastore_table_row_count() call failed")
+    }
+
+    /// Iterate over all rows in the TX state,
+    /// i.e. committed_state ∪ insert_table ∖ delete_table.
+    #[inline]
+    fn iter(&self) -> impl Iterator<Item = Self::Row> {
+        let table_id = Self::table_id();
+        let iter = sys::datastore_table_scan_bsatn(table_id).expect("datastore_table_scan_bsatn() call failed");
+        TableIter::new(iter)
+    }
+
+    /// Inserts `row` into the TX state,
+    /// i.e. removes it from the delete table or adds it to the insert table as appropriate.
+    ///
+    /// The return value is the inserted row, with any auto-incrementing columns replaced with computed values.
+    /// The `insert` method always returns the inserted row,
+    /// even when the table contains no auto-incrementing columns.
+    ///
+    /// May panic if inserting the row violates any constraints.
+    /// Callers which intend to handle constraint violation errors should instead use [`Self::try_insert`].
+    ///
+    /// Note that, in languages where error handling is based on exceptions,
+    /// no distinction is provided between `Table::insert` and `Table::try_insert`.
+    /// A single method `insert` is defined which throws an exception on failure,
+    /// and callers may either install handlers around it or allow the exception to bubble up.
+    ///
+    /// Note on MVCC: because callers have no way to determine if the row was previously present,
+    /// two concurrent transactions which delete the same row
+    /// may be ordered arbitrarily with respect to one another
+    /// while maintaining sequential consistency, assuming no other conflicts.
+    #[track_caller]
+    fn insert(&self, row: Self::Row) -> Self::Row {
+        self.try_insert(row).unwrap_or_else(|e| panic!("{e}"))
+    }
+
+    /// The error type for this table for unique constraint violations. Will either be
+    /// [`UniqueConstraintViolation`] if the table has any unique constraints, or [`Infallible`]
+    /// otherwise.
+    type UniqueConstraintViolation: MaybeError<UniqueConstraintViolation>;
+    /// The error type for this table for auto-increment overflows. Will either be
+    /// [`AutoIncOverflow`] if the table has any auto-incrementing columns, or [`Infallible`]
+    /// otherwise.
+    type AutoIncOverflow: MaybeError<AutoIncOverflow>;
```

Comment:
Again, not how the `Table` trait is specified.

Reply:
I couldn't tell how it was supposed to be from the proposal - "For tables with constraints, [TryInsertError] is an enum generated by the `#[spacetimedb::table]` macro. [...] TODO: specify a generated name for this type?". In the interest of generating as few identifiers as possible, I made it so that it would be an enum defined in the bindings crate, but with variants defined based on the given `Table`. Are you amenable to using this implementation? IMO it's a better UX & better docs than the generated enum version mentioned in the proposal.

Reply:
I am amenable to this. Leave a comment on the proposal and I'll update it to reflect your version.